### PR TITLE
feat: run tests with sanitizers (tsan, asan)

### PR DESCRIPTION
**Pull request by punkeel**
Ref: #293

### CODE REVIEW

File: .github/workflows/asan.yml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    timeout-minutes: 120
+    strategy:
+      fail-fast: false
+      # Always run main branch builds to completion. This allows the cache to
+      # stay mostly up-to-date in situations where a single job fails due to
+      # e.g. a flaky test.
+      # Don't fast-fail on tag build because publishing binaries shouldn't be
+      # prevented if 'cargo publish' fails (which can be a false negative).
+      # fail-fast:
+      #  ${{ github.event_name == 'pull_request' || (github.ref !=
```
Comment (punkeel) [inline .github/workflows/asan.yml:21]: TODO: bring back fail-fast

**Comment by punkeel** [comment]
**Current status**: it takes 1h per sanitizer run, they run in parallel (can't enable multiple sanitizers in one go)
ASAN builds & runs just fine, MSAN/TSAN still need work.
If you'd like, we can merge the ASAN part, while the rest is being worked on.

Not my decision to make: it may be interesting to publish ASAN-enabled builds for consumption by other projects (say: deno).  I'll keep this work out of the current PR.


--- 
*asan*: builds and runs tests just fine. Yay.

---
*Thread san* doesn't build, I suspect V8 + Rust include the same tsan library with the same function definitions. Either that, or libc++ is embedded in the v8 library _and_ in the rusty_v8 library "again".
I'm not sure how to fix it, but I'll dig a bit.
```
= note: /usr/bin/ld: /home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/gn_out/obj//librusty_v8.a(cxa_guard.o): in function `__cxa_guard_abort':
          ./../../../../buildtools/third_party/libc++abi/trunk/src/cxa_guard.cpp:47: multiple definition of `__cxa_guard_abort'; /usr/share/rust/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc-nightly_rt.tsan.a(tsan_interceptors_posix.cpp.o):/rustc/llvm/src/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:882: first defined here
```

---
*Memory san* run: https://github.com/denoland/rusty_v8/runs/3041645518#step:9:7645
```
    Finished test [unoptimized + debuginfo] target(s) in 49m 13s
     Running `/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3`

running 12 tests
test array_buffer::test_default_allocator ... ok
test array_buffer::test_rust_allocator ... ok
test inspector::string_view_display ... ok
==13097==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55e0ee76d89e  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x1c989e)
    #1 0x55e0ee80c6a8  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x2686a8)
    #2 0x55e0ee80d8e9  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x2698e9)
    #3 0x55e0ee7cc470  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x228470)
    #4 0x55e0ee7be15c  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x21a15c)
    #5 0x55e0f068ebbf  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x20eabbf)
    #6 0x55e0f06ce6e3  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x212a6e3)
    #7 0x55e0f06cb3e2  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x21273e2)
    #8 0x55e0f068e754  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x20ea754)
    #9 0x55e0f06aca6b  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x2108a6b)
    #10 0x55e0f0743164  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x219f164)
    #11 0x55e0f08001f8  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x225c1f8)
    #12 0x55e0f08227db  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x227e7db)
    #13 0x55e0f07ffbd6  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x225bbd6)
    #14 0x55e0f07465f3  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x21a25f3)
    #15 0x55e0f06ceec3  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x212aec3)
    #16 0x55e0f06cdf19  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x2129f19)
    #17 0x55e0f06cbf1e  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x2127f1e)
    #18 0x55e0f06caeae  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x2126eae)
    #19 0x55e0f06c1f03  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x211df03)
    #20 0x55e0f06677c4  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x20c37c4)
    #21 0x55e0f06b9611  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x2115611)
    #22 0x55e0f06ba3da  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x21163da)
    #23 0x55e0ee7775ea  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x1d35ea)
    #24 0x55e0ee7be77f  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x21a77f)
    #25 0x55e0ee73b598  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x197598)
    #26 0x55e0ee766b99  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x1c2b99)
    #27 0x55e0f0b39e0e  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x2595e0e)
    #28 0x55e0f0b959d0  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x25f19d0)
    #29 0x55e0f0ba17fb  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x25fd7fb)
    #30 0x55e0f0b94385  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x25f0385)
    #31 0x55e0f0b90095  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x25ec095)
    #32 0x55e0f0cbe8dc  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x271a8dc)
    #33 0x55e0f0b95772  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x25f1772)
    #34 0x55e0f0ba17fb  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x25fd7fb)
    #35 0x55e0f0b94ed5  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x25f0ed5)
    #36 0x55e0f0b8ff55  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x25ebf55)
    #37 0x55e0f0cbe2a1  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x271a2a1)
    #38 0x55e0ee766a7c  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x1c2a7c)
    #39 0x55e0ee777647  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x1d3647)
    #40 0x7fdc8838b0b2  (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #41 0x55e0ee6dda5d  (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x139a5d)

SUMMARY: MemorySanitizer: use-of-uninitialized-value (/home/runner/work/rusty_v8/rusty_v8/target/x86_64-unknown-linux-gnu/debug/deps/rusty_v8-85f696c789063cc3+0x1c989e) 
Exiting
error: test failed, to rerun pass '--lib'
test inspector::tests::test_channel ... 
```

**Comment by bartlomieju** [comment]
@punkeel thank you very much! This work is much appreciated, as getting `rusty_v8` building against those sanitizers had been a long standing goal of this project. 

> address san: builds and runs tests just fine. Yay.
Note: the leak detector is disabled. Might be worth investigating… or not. I'm not sure cargo testsare designed with leak-prevention in mind.

Merging only ASAN build sounds good to me

> memory san run: https://github.com/denoland/rusty_v8/runs/3041645518#step:9:7645

> Note it was able to run a few tests before crashing, which is better than crashing before running anything… but could indicate a real problem 😕

I'd say it's quite probable there's actual bug there so great to see it uncovered!

I'll make sure to discuss this PR with the rest of the team ASAP.

File: .github/workflows/asan.yml (hunk 1 above)
Comment (punkeel) [reply .github/workflows/asan.yml]: (or not? Is there value running all sanitizers for all PRs, even if one fails?)

**Comment by punkeel** [comment]
(PR still WIP, experimenting with ubsan&co. I amended my comment to fix some wrong things I said)

**Comment by punkeel** [comment]
Current status:
- Really not sure I can get tsan working in a reasonable time, due to the static(?) linking of libc++
- Symbols are printed in the reports, thanks llvm-symbolizer
- Now running cargo test in release mode. Considering running other simple scripts as well, ideas welcome on that front.
- as of now: everything is broken, yay :) I suspect I misconfigured CI, I’ll get it fixed.
- Asan worked at least a few times, with leak detector disabled. I’m trying to enable the leak detector again.
- ubsan (supported by V8 but not Rust) can be enabled alongside asan. I’ll see if I can get it to work.

**Pull request closed without merging**